### ISSUE TITLE
ambient precedence of operand of `--` and `++`

### ISSUE THREAD

**Issue by zygoloid** [carbon-language/carbon-lang#3079]
We say that `++` and `--` can only be used as a statement, but what kinds of expression can appear within them? For example, is:

```
++x + y;
```

... parsed as `++(x + y);`, or is it a syntactic error?

There aren't many expression forms that can be reference expressions in Carbon's syntax -- only function calls, `a.b`, `a->b`, `*p`, `a[i]`, and `if` expressions can be reference expressions at the moment. All other operators always result in value expressions or initializing expressions. So we'll still always diagnose things like `++x + y;` due to an expression category error, but the diagnostic may not be clear about the nature of the problem.

On the other hand, if we impose a precendence restriction (eg, we could say the operand must have the precedence of `*p` or higher) then we would disallow

```
++if cond then x else y;
```

... which might be a desirable operation.

**Comment by chandlerc** [comment]
Not just a reference expression, but a reference expression that isn't ephemeral. I think that even excludes function calls. And I can't think of a real use case for incrementing a function call.

FWIW, I'm already a bit hesitant about `if` expressions being (durable) reference expressions. And I think the syntax above isn't particularly readable. I'd personally prefer paretheses there.

I'd similarly prefer parentheses on the LHS of assignment: `if cond then x else y = 42;` seems way too hard to read. But `(if cond then x else y) = 42;` seems better. (Although again, I'm still not a huge fan. But maybe that means I'm a bit biased and about the syntax here.)

**Comment by zygoloid** [comment]
I suggest we put the assignment operators here in the precedence diagram:

```mermaid
%%{init: {'themeVariables': {'fontFamily': 'monospace'}}}%%
graph BT
    top(("..."))

    pointer{"*x<br>
             &x"}
    click pointer "https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/expressions/pointer.md"

    negation["-x"]
    click negation "https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/expressions/arithmetic.md"

    complement["^x"]
    click complement "https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/expressions/bitwise.md"

    incDec["++x;<br>--x;"]
    click incDec "https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/assignment.md"

    style incDec color:#0f0,stroke:#0b0,fill:#020

    unary(("..."))

    pointer --> top
    negation --> pointer
    complement --> pointer
    incDec --> pointer
    unary --> negation & complement
```

```mermaid
%%{init: {'themeVariables': {'fontFamily': 'monospace'}}}%%
graph BT
    logicalOperand(("..."))

    and>"x and y"]
    click and "https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/expressions/logical_operators.md"

    or>"x or y"]
    click or "https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/expressions/logical_operators.md"

    logicalExpression((" "))

    if>"if x then y else z"]
    click if "https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/expressions/if.md"

    insideParens["(...)"]

    assignment["x = y;<br>x $= y;"]
    click assignment "https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/assignment.md"

    style assignment color:#0f0,stroke:#0b0,fill:#020

    expressionStatement["x;"]

    and & or --> logicalOperand
    logicalExpression --> and & or
    if & assignment & expressionStatement --> logicalExpression
    insideParens --> if
```

That'd mean that we require all the parentheses in the following example:

```carbon
(if c then a else b) = (if c then 1 else 2);
++(if c then a else b);
```

... but do not require any parentheses in:

```carbon
*p = a <= b * 2 + 1 and not c;
*p = 5 as i32;
```

We could put increment / decrement operators in the same place as simple / compound assignment, but that'd mean that

```carbon
++a + 1 <= 3 or not b;
```

... would parse as ...

```carbon
++(((a + 1) <= 3) or (not b));
```

... which would likely result in worse diagnostics, though it'd be rejected either way.

**Comment by zygoloid** [comment]
From discussion with @chandlerc: we want the rules for initialization and assignment to match, so we want unparenthesized `if` to be able to appear on the right of an assignment. We could have different rules for the right hand side and the left hand side, but:

- we don't do that anywhere else currently (and doing so would complicate our precedence rules), and
- we don't actually need to: an unparenthesized `if` on the left hand side would be invalid for (at least) two other reasons: it would be interpreted as an `if` statement instead, and could never be a reference expression anyway

So from a syntactic perspective, the status quo (any expression can be an operand of assignments) is actually what we want.

For increment and decrement, we do like having higher precedence, but this actually isn't a *change*, because the cases we're eliminating with the precedence rules are already eliminated by the expression category rules, so this is just changing how we present  the rules, not which programs are valid.

In conclusion: we want assignment at the bottom of the precedence diagram (below `if`), and increment / decrement immediately below dereference, but we think this is *not* a change from the approved rules, only a change in the design documentation.
